### PR TITLE
[2FA] Add a lot of log-points

### PR DESCRIPTION
**Pull request by nupplaphil**
Maybe help getting the bug at #11712 

Because I don't get any error during my tests :(

### CODE REVIEW

**Review by MrPetovan — APPROVED** [review]

**Comment by MrPetovan** [comment]
This error happens when the session expires so it's likely you weren't able to test this case personally.

**Comment by nupplaphil** [comment]
I really don't like the concept of throwing exceptions to redirect to other pages .. It's easy to oversee it and accidentally catch it .. Like I did at first ;-)

**Review by MrPetovan — APPROVED** [review]

**Comment by MrPetovan** [comment]
So you found the issue?

**Comment by MrPetovan** [comment]
> I really don't like the concept of throwing exceptions to redirect to other pages .. It's easy to oversee it and accidentally catch it .. Like I did at first ;-)

I agree, but in this particular case, it still is better than an `exit()` you can't do anything about.

File: src/Module/Security/TwoFactor/SignOut.php [hunk 1]
```diff
@@ -108,7 +108,7 @@ protected function content(array $request = []): string
 				info($this->t('Logged out.'));
 				$this->baseUrl->redirect();
 			}
-		} catch (NotFoundException $exception) {
+		} catch (TwoFactor\Exception\TrustedBrowserNotFoundException $exception) {
```
Comment (Quix0r) [inline src/Module/Security/TwoFactor/SignOut.php:111]: Can you import with `use` it and use here the short version?

File: src/Module/Security/TwoFactor/Trust.php [hunk 2]
```diff
@@ -103,13 +117,17 @@ protected function content(array $request = []): string
...
 		if ($this->cookie->get('2fa_cookie_hash')) {
 			try {
-				$trustedBrowser = $this->trustedBrowserRepositoy->selectOneByHash($this->cookie->get('2fa_cookie_hash'));
+				$trustedBrowser = $this->trustedBrowserRepository->selectOneByHash($this->cookie->get('2fa_cookie_hash'));
 				if (!$trustedBrowser->trusted) {
 					$this->auth->setForUser($this->app, User::getById($this->app->getLoggedInUserId()), true, true);
 					$this->baseUrl->redirect();
 				}
-			} catch (NotFoundException $exception) {
+			} catch (TwoFactor\Exception\TrustedBrowserNotFoundException $exception) {
```
Comment (Quix0r) [inline src/Module/Security/TwoFactor/Trust.php:125]: Same here, please use `use` instead.

File: src/Module/Security/TwoFactor/Trust.php [hunk 3]
```diff
@@ -103,13 +117,17 @@ protected function content(array $request = []): string
...
 				$this->logger->notice('Trusted Browser of the cookie not found.', ['cookie_hash' => $this->cookie->get('trusted'), 'uid' => $this->app->getLoggedInUserId(), 'exception' => $exception]);
+			} catch (TwoFactor\Exception\TrustedBrowserPersistenceException $exception) {
```
Comment (Quix0r) [inline src/Module/Security/TwoFactor/Trust.php:127]: Same here, please use `use` instead.

File: src/Security/TwoFactor/Exception/TrustedBrowserNotFoundException.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ */
+
+namespace Friendica\Security\TwoFactor\Exception;
+
+use Exception;
+
+class TrustedBrowserNotFoundException extends \RuntimeException
+{
+	public function __construct($message = '', Exception $previous = null)
```
Comment (Quix0r) [inline src/Security/TwoFactor/Exception/TrustedBrowserNotFoundException.php:28]: Please add type-hints as the super (parent) constructor has.

File: src/Security/TwoFactor/Exception/TrustedBrowserPersistenceException.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ */
+
+namespace Friendica\Security\TwoFactor\Exception;
+
+use Throwable;
+
+class TrustedBrowserPersistenceException extends \RuntimeException
+{
+	public function __construct($message = "", Throwable $previous = null)
```
Comment (Quix0r) [inline src/Security/TwoFactor/Exception/TrustedBrowserPersistenceException.php:28]: Please add type-hints as the super (parent) constructor has.

File: src/Security/TwoFactor/Repository/TrustedBrowser.php [hunk 6]
```diff
@@ -35,83 +33,122 @@ class TrustedBrowser
...
 	{
-		$fields = $this->db->selectFirst(self::$table_name, [], ['cookie_hash' => $cookie_hash]);
+		try {
+			$fields = $this->db->selectFirst(self::$table_name, [], ['cookie_hash' => $cookie_hash]);
+		} catch (\Exception $exception) {
+			throw new TwoFactor\Exception\TrustedBrowserPersistenceException(sprintf('Internal server error when retrieving cookie hash \'%s\'', $cookie_hash));
+		}
 		if (!$this->db->isResult($fields)) {
-			throw new NotFoundException('');
+			throw new TwoFactor\Exception\TrustedBrowserNotFoundException(sprintf('Cookie hash \'%s\' not found', $cookie_hash));
```
Comment (Quix0r) [inline src/Security/TwoFactor/Repository/TrustedBrowser.php:64]: Please use `use` instead.

File: src/Security/TwoFactor/Repository/TrustedBrowser.php [hunk 7]
```diff
@@ -35,83 +33,122 @@ class TrustedBrowser
...
-		}
+			$trustedBrowsers = [];
+			foreach ($rows as $fields) {
+				$trustedBrowsers[] = $this->factory->createFromTableRow($fields);
+			}
+			return new TwoFactor\Collection\TrustedBrowsers($trustedBrowsers);
 
-		return new TrustedBrowsers($trustedBrowsers);
+		} catch (\Exception $exception) {
+			throw new TwoFactor\Exception\TrustedBrowserPersistenceException(sprintf('selection for uid \'%s\' wasn\'t successful.', $uid));
```
Comment (Quix0r) [inline src/Security/TwoFactor/Repository/TrustedBrowser.php:89]: Please use `use` instead.

File: src/Security/TwoFactor/Repository/TrustedBrowser.php [hunk 8]
```diff
@@ -35,83 +33,122 @@ class TrustedBrowser
...
+	 * @throws TwoFactor\Exception\TrustedBrowserPersistenceException
 	 */
-	public function save(Model\TrustedBrowser $trustedBrowser): bool
+	public function save(TwoFactor\Model\TrustedBrowser $trustedBrowser): bool
 	{
-		return $this->db->insert(self::$table_name, $trustedBrowser->toArray(), $this->db::INSERT_UPDATE);
+		try {
+			return $this->db->insert(self::$table_name, $trustedBrowser->toArray(), $this->db::INSERT_UPDATE);
+		} catch (\Exception $exception) {
+			throw new TwoFactor\Exception\TrustedBrowserPersistenceException(sprintf('Couldn\'t save trusted Browser with cookie_hash \'%s\'', $trustedBrowser->cookie_hash));
```
Comment (Quix0r) [inline src/Security/TwoFactor/Repository/TrustedBrowser.php:105]: Please use `use` instead.

File: src/Security/TwoFactor/Repository/TrustedBrowser.php [hunk 9]
```diff
@@ -35,83 +33,122 @@ class TrustedBrowser
...
+	 * @throws TwoFactor\Exception\TrustedBrowserPersistenceException
 	 */
-	public function remove(Model\TrustedBrowser $trustedBrowser): bool
+	public function remove(TwoFactor\Model\TrustedBrowser $trustedBrowser): bool
 	{
-		return $this->db->delete(self::$table_name, ['cookie_hash' => $trustedBrowser->cookie_hash]);
+		try {
+			return $this->db->delete(self::$table_name, ['cookie_hash' => $trustedBrowser->cookie_hash]);
+		} catch (\Exception $exception) {
+			throw new TwoFactor\Exception\TrustedBrowserPersistenceException(sprintf('Couldn\'t delete trusted Browser with cookie hash \'%s\'', $trustedBrowser->cookie_hash));
```
Comment (Quix0r) [inline src/Security/TwoFactor/Repository/TrustedBrowser.php:121]: Please use `use` instead.

File: src/Security/TwoFactor/Repository/TrustedBrowser.php [hunk 10]
```diff
@@ -35,83 +33,122 @@ class TrustedBrowser
...
+	 *
+	 * @throws TwoFactor\Exception\TrustedBrowserPersistenceException
 	 */
 	public function removeForUser(int $local_user, string $cookie_hash): bool
 	{
-		return $this->db->delete(self::$table_name, ['cookie_hash' => $cookie_hash,'uid' => $local_user]);
+		try {
+			return $this->db->delete(self::$table_name, ['cookie_hash' => $cookie_hash, 'uid' => $local_user]);
+		} catch (\Exception $exception) {
+			throw new TwoFactor\Exception\TrustedBrowserPersistenceException(sprintf('Couldn\'t delete trusted Browser for user \'%s\' and cookie hash \'%s\'', $local_user, $cookie_hash));
```
Comment (Quix0r) [inline src/Security/TwoFactor/Repository/TrustedBrowser.php:138]: Please use `use` instead.

File: src/Security/TwoFactor/Repository/TrustedBrowser.php [hunk 11]
```diff
@@ -35,83 +33,122 @@ class TrustedBrowser
...
 	 * @return bool
-	 * @throws \Exception
 	 */
 	public function removeAllForUser(int $local_user): bool
 	{
-		return $this->db->delete(self::$table_name, ['uid' => $local_user]);
+		try {
+			return $this->db->delete(self::$table_name, ['uid' => $local_user]);
+		} catch (\Exception $exception) {
+			throw new TwoFactor\Exception\TrustedBrowserPersistenceException(sprintf('Couldn\'t delete trusted Browsers for user \'%s\'', $local_user));
```
Comment (Quix0r) [inline src/Security/TwoFactor/Repository/TrustedBrowser.php:151]: Please use `use` instead.

**Review by Quix0r — CHANGES_REQUESTED** [review]

**Pull request merged**